onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h

```diff
@@ -200,104 +206,131 @@ Status TreeEnsembleCommon<InputType, ThresholdType, OutputType>::Init(
     }
-    idi.insert(std::pair<TreeNodeElementId, TreeNodeElement<ThresholdType>*>(node.id, &node));
+    idi.insert(std::pair<TreeNodeElementId, uint32_t>(node_tree_id, i));
   }
 
   TreeNodeElementId coor;
+  i = 0;
   for (auto it = nodes_.begin(); it != nodes_.end(); ++it, ++i) {
-    if (!it->is_not_leaf)
+    if (!it->is_not_leaf())
       continue;
-    i = std::distance(nodes_.begin(), it);
-    coor.tree_id = it->id.tree_id;
+
+    TreeNodeElementId& node_tree_id = node_tree_ids[i];
+    coor.tree_id = node_tree_id.tree_id;
     coor.node_id = static_cast<int>(nodes_truenodeids[i]);
 
     auto found = idi.find(coor);
     if (found == idi.end()) {
       ORT_THROW("Unable to find node ", coor.tree_id, "-", coor.node_id, " (truenode).");
     }
-    if (coor.node_id >= 0 && coor.node_id < n_nodes_) {
-      it->truenode = found->second;
-      if ((it->truenode->id.tree_id != it->id.tree_id) ||
-          (it->truenode->id.node_id == it->id.node_id)) {
-        ORT_THROW("One falsenode is pointing either to itself, either to another tree.");
-      }
-    } else
-      it->truenode = nullptr;
+    truenode_ids[i] = (coor.node_id >= 0 && coor.node_id < n_nodes_) ? found->second : 0;
 
     coor.node_id = static_cast<int>(nodes_falsenodeids[i]);
     found = idi.find(coor);
     if (found == idi.end()) {
       ORT_THROW("Unable to find node ", coor.tree_id, "-", coor.node_id, " (falsenode).");
     }
-    if (coor.node_id >= 0 && coor.node_id < n_nodes_) {
-      it->falsenode = found->second;
-      if ((it->falsenode->id.tree_id != it->id.tree_id) ||
-          (it->falsenode->id.node_id == it->id.node_id)) {
-        ORT_THROW("One falsenode is pointing either to itself, either to another tree.");
-      }
-    } else
-      it->falsenode = nullptr;
+    falsenode_ids[i] = (coor.node_id >= 0 && coor.node_id < n_nodes_) ? found->second : 0;
   }
 
-  int64_t previous = -1;
-  for (i = 0; i < static_cast<size_t>(n_nodes_); ++i) {
-    if ((previous == -1) || (previous != nodes_[i].id.tree_id))
-      roots_.push_back(&(nodes_[i]));
-    previous = nodes_[i].id.tree_id;
+  // sort targets
+  std::vector<std::pair<TreeNodeElementId, uint32_t>> indices;
+  indices.reserve(target_class_nodeids.size());
+  for (i = 0, limit = target_class_nodeids.size(); i < limit; i++) {
+    indices.emplace_back(std::pair<TreeNodeElementId, uint32_t>(
+        TreeNodeElementId(target_class_treeids[i], target_class_nodeids[i]),
+        i));
   }
+  std::sort(indices.begin(), indices.end());
 
+  // Initialize the leaves.
   TreeNodeElementId ind;
   SparseValue<ThresholdType> w;
-  for (i = 0, limit = target_class_nodeids.size(); i < limit; i++) {
-    ind.tree_id = static_cast<int>(target_class_treeids[i]);
-    ind.node_id = static_cast<int>(target_class_nodeids[i]);
+  size_t indi;
+  for (indi = 0, limit = target_class_nodeids.size(); indi < limit; ++indi) {
+    ind = indices[indi].first;
```

Comment:
>indi [](http://example.com/codeflow?start=18&length=4)

Does this unintentionally create an entry that does not exist?

Reply:
an entry in which container?